File: 04/solution.ts
const input = await Bun.file(`${import.meta.dir}/input.txt`).text();
const lines = input.trim().split("\n");

type Point = { x: number; y: number };

function neighbors(grid: string[][], { x, y }: Point) {
	return [
		{ x: x - 1, y },
		{ x: x + 1, y },
		{ x: x - 1, y: y + 1 },
		{ x: x + 1, y: y + 1 },
		{ x: x - 1, y: y - 1 },
		{ x: x + 1, y: y - 1 },
		{ x: x, y: y + 1 },
		{ x: x, y: y - 1 },
	].filter(
		(p) => p.x >= 0 && p.x < grid.length && p.y >= 0 && p.y < grid[0].length,
	);
}

function removeRolls(grid: string[][], points: Point[]) {
	const newGrid = structuredClone(grid);

	for (const p of points) {
		newGrid[p.x][p.y] = ".";
	}

	return newGrid;
}

function part1() {
	let sum = 0;
	const grid = lines.map((l) => l.split(""));

	for (let y = 0; y < grid.length; y++) {
		for (let x = 0; x < grid[0].length; x++) {
			const cell = grid[x][y];
			if (cell === "@") {
				const rolls = neighbors(grid, { x, y })
					.map((n) => grid[n.x][n.y])
					.filter((x) => x === "@").length;

				if (rolls < 4) {
					sum++;
				}
			}
		}
	}

	return sum;
}

function part2() {
	let sum = 0;
	let grid = lines.map((l) => l.split(""));
	let stillRemoving = true;
	let accessible: Point[] = [];

	while (stillRemoving) {
		for (let y = 0; y < grid.length; y++) {
			for (let x = 0; x < grid[0].length; x++) {
				const cell = grid[x][y];
				if (cell === "@") {
					const rolls = neighbors(grid, { x, y })
						.map((n) => grid[n.x][n.y])
						.filter((x) => x === "@").length;

					if (rolls < 4) {
						accessible.push({ x, y });
					}
				}
			}
		}

		if (accessible.length === 0) stillRemoving = false;
		if (accessible.length > 0) {
			sum += accessible.length;
			grid = removeRolls(grid, accessible);
			accessible = [];
		}
	}

	return sum;
}

console.log("Part 1:", part1());
console.log("Part 2:", part2());

File: 04/solution.ml
open Base
open Stdio

let read_input () = In_channel.read_lines "04/input.txt"

let neighbors grid (x, y) =
  let rows = Array.length grid in
  let cols = Array.length grid.(0) in
  [
    (x - 1, y);
    (x + 1, y);
    (x, y - 1);
    (x, y + 1);
    (x - 1, y - 1);
    (x - 1, y + 1);
    (x + 1, y - 1);
    (x + 1, y + 1);
  ]
  |> List.filter ~f:(fun (x, y) -> x >= 0 && x < rows && y >= 0 && y < cols)

let part1 lines =
  let grid = lines |> List.map ~f:String.to_array |> List.to_array in
  let rows = Array.length grid in
  let cols = Array.length grid.(0) in
  List.range 0 rows
  |> List.sum
       (module Int)
       ~f:(fun x ->
         List.range 0 cols
         |> List.count ~f:(fun y ->
             Char.(grid.(x).(y) = '@')
             && neighbors grid (x, y)
                |> List.count ~f:(fun (nx, ny) -> Char.(grid.(nx).(ny) = '@'))
                |> fun n -> n < 4))

let part2 lines =
  let grid = lines |> List.map ~f:String.to_array |> List.to_array in
  let rows = Array.length grid in
  let cols = Array.length grid.(0) in
  let rec loop sum =
    let accessible =
      List.range 0 rows
      |> List.concat_map ~f:(fun x ->
          List.range 0 cols
          |> List.filter_map ~f:(fun y ->
              if
                Char.(grid.(x).(y) = '@')
                && neighbors grid (x, y)
                   |> List.count ~f:(fun (nx, ny) ->
                       Char.(grid.(nx).(ny) = '@'))
                   |> fun n -> n < 4
              then Some (x, y)
              else None))
    in
    match accessible with
    | [] -> sum
    | pts ->
        List.iter pts ~f:(fun (x, y) -> grid.(x).(y) <- '.');
        loop (sum + List.length pts)
  in
  loop 0

let () =
  let lines = read_input () in
  printf "Part 1: %d\n" (part1 lines);
  printf "Part 2: %d\n" (part2 lines)
